Change "Override Execution Context" button to "Record" button on nodes

somebody1234 opened the pull request:
### Pull Request Description
- Close #9164
  - Fix appearance of Record/Record Once icon in top menu
  - Change icon for overriding execution context to record icon
  - Unconditionally show per-node record icon if it is set
  - Remove the ability to override the execution context to disabled
  - Fix the icon for nodes with an overridden execution context always being the Enso icon

### Important Notes
None

### Screenshot
![image](https://github.com/enso-org/enso/assets/4046547/40d0fa4b-415e-410c-8a8c-e2487aaf838e)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
Use new icon for overridding output context; prevent overriding the output context to disabled

somebody1234 committed:
Always show button to disable recording if recording is enabled

somebody1234 committed:
Fix appearance of Record/Record Once button in top bar

somebody1234 committed:
Fix bugs

somebody1234 committed:
WIP: does it work?

somebody1234 commented:
@farmaazon you mentioned you might have wanted to take a look at this

jdunkerley commented on app/gui2/src/components/CircularMenu.vue:
```diff
@@ -4,13 +4,13 @@ import ToggleIcon from '@/components/ToggleIcon.vue'
 
 const props = defineProps<{
   isOutputContextEnabledGlobally: boolean
-  isOutputContextOverridden: boolean
+  isRecordingOverridden: boolean
```
shouldn't we be consistent in naming both these properties?

somebody1234 replied:
ah, good catch, thanks

somebody1234 committed:
Rename `OutputContext` to `Recording`

somebody1234 committed:
WIP: Attempt to fix

somebody1234 committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

somebody1234 committed:
Minor fixes

somebody1234 committed:
Fix icon for nodes that are recording

Primary subject is still wrong, and toggling recording may still be
broken

somebody1234 committed:
Fix self argument detection

somebody1234 committed:
Remove debug log

somebody1234 commented:
@AdRiley just a heads up: i've decided to remove support for the old behavior, since it doesn't really work well with the refactors.
i figure it should be simple enough to grab the old code from an old commit anyway

somebody1234 committed:
Attempt (and fail) to fix bug

somebody1234 committed:
Fix obsolete FIXME

somebody1234 commented on app/gui2/src/util/ast/node.ts:
```diff
@@ -1,19 +1,28 @@
   if (!nodeCode) return
   const pattern = nodeCode instanceof Ast.Assignment ? nodeCode.pattern : undefined
   const rootSpan = nodeCode instanceof Ast.Assignment ? nodeCode.expression : nodeCode
+  const { innerExpr, matches } = prefixes.extractMatches(rootSpan)
```
@kazcw would you happen to know why this errors with this error on the second click?
```
Uncaught Error: Assertion failed: id in module
    bail assert.ts:68
    assert assert.ts:6
    get mutableModule.ts:253
    children tree.ts:146
    zipLongest iterable.ts:54
    matchSubtree match.ts:75
    match match.ts:27
    matches prefixes.ts:30
    extractMatches prefixes.ts:29
    nodeFromAst node.ts:16
    readFunctionAst graphDatabase.ts:347
    updateState index.ts:146
    handleModuleUpdate index.ts:127
    handle index.ts:94
```

somebody1234 replied:
or perhaps it's erroring on `MutableModule`s only? perhaps due to some inconsistency in the state of the tree?
(edit: it seems like the initial module is a `MutableModule` and *that* has no problems...)

also worth noting the edit seems to succeed, so it's probably erroring during the redrawing of the new tree

kazcw replied:
The problem is caused by mutating data returned from Y.Js. The current `develop` changes access patterns and happens to fix the case encountered in this PR. I'll open a PR soon addressing the root cause.

somebody1234 committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

farmaazon commented on app/gui2/src/assets/icons.svg:
```diff
@@ -709,7 +711,7 @@
 
   <symbol id="scissors" viewBox="0 0 16 16" width="16" height="16" fill="none">
-    <path fill-rule="evenodd" clip-rule="evenodd" d="M13.7699 12.8559C14.5352 13.2977 15.5137 13.0355 15.9555 12.2703C16.066 12.0789 16.0005 11.8343 15.8091 11.7238L15.4786 11.533L10.3136 8.55098L8.31356 9.70568L13.4394 12.6651L13.444 12.6677L13.7699 12.8559ZM5.76183 10.0243L4.9631 10.4855C5.07499 10.6039 5.17484 10.7358 5.26034 10.8807C5.96159 12.0689 5.43125 13.6806 4.0758 14.4806C2.72034 15.28 ...
+    <path fill-rule="evenodd" clip-rule="evenodd" d="M13.7699 12.8559C14.5352 13.2977 15.5137 13.0355 15.9555 12.2703C16.066 12.0789 16.0004 11.8343 15.8091 11.7238L15.4786 11.533L10.3136 8.55098L8.31355 9.70568L13.4394 12.6651L13.444 12.6677L13.7699 12.8559ZM5.76182 10.0243L4.96309 10.4855C5.07499 10.6039 5.17484 10.7358 5.26033 10.8807C5.96158 12.0689 5.43125 13.6806 4.07579 14.4806C2.72033 15. ...
```
I assume these changes are the new export figma?

somebody1234 replied:
yup

farmaazon commented on app/gui2/src/stores/graph/graphDatabase.ts:
```diff
@@ -458,8 +464,10 @@ export interface Node {
   outerExprId: Ast.AstId
   pattern: Ast.Ast | undefined
   rootSpan: Ast.Ast
+  innerExpr: Ast.Ast
```
We definitely need documentation here, describing what exactly is `outerExprId`, `root expression` and `inner expression`. And rename `rootSpan` to `rootExpr` for consistency.

somebody1234 committed:
Address review

farmaazon commented:
I guess it also fixes [#7976](https://github.com/enso-org/enso/issues/7976)?

somebody1234 commented:
probably, yeah. tests haven't been added in this PR though, maybe that could be split out into another task?

somebody1234 commented:
anyone know what's going on with the tests?
https://github.com/enso-org/enso/actions/runs/8094866300/job/22120440156?pr=9188#step:8:466

```
  INFO ide_ci::program::command: npm⚠️ ⎯⎯⎯⎯⎯⎯ Failed Suites 8 ⎯⎯⎯⎯⎯⎯⎯
 INFO ide_ci::program::command: npmℹ️ 
 INFO ide_ci::program::command: npm⚠️ 
 INFO ide_ci::program::command: npm⚠️  FAIL  src/composables/__tests__/selection.test.ts [ src/composables/__tests__/selection.test.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/providers/__tests__/widgetRegistry.test.ts [ src/providers/__tests__/widgetRegistry.test.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/components/ComponentBrowser/__tests__/input.test.ts [ src/components/ComponentBrowser/__tests__/input.test.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/components/GraphEditor/__tests__/dragging.test.ts [ src/components/GraphEditor/__tests__/dragging.test.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/stores/graph/__tests__/graphDatabase.test.ts [ src/stores/graph/__tests__/graphDatabase.test.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/composables/selection.ts [ src/composables/selection.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/components/GraphEditor/collapsing.ts [ src/components/GraphEditor/collapsing.ts ]
 INFO ide_ci::program::command: npm⚠️  FAIL  src/util/ast/node.ts [ src/util/ast/node.ts ]
 INFO ide_ci::program::command: npm⚠️ TypeError: Cannot read properties of undefined (reading '__wbindgen_add_to_stack_pointer')
 INFO ide_ci::program::command: npm⚠️  ❯ Module.parse rust-ffi/pkg/rust_ffi.js:156:14
 INFO ide_ci::program::command: npm⚠️     154|         return v2;
 INFO ide_ci::program::command: npm⚠️     155|     } finally {
 INFO ide_ci::program::command: npm⚠️     156|         wasm.__wbindgen_add_to_stack_pointer(16);
 INFO ide_ci::program::command: npm⚠️        |              ^
 INFO ide_ci::program::command: npm⚠️     157|     }
 INFO ide_ci::program::command: npm⚠️     158| }
 INFO ide_ci::program::command: npm⚠️  ❯ parseEnso shared/ast/parse.ts:65:16
 INFO ide_ci::program::command: npm⚠️  ❯ parseBlockWithSpans shared/ast/parse.ts:541:16
 INFO ide_ci::program::command: npm���️  ❯ parseBlock shared/ast/parse.ts:521:10
 INFO ide_ci::program::command: npm⚠️  ❯ Module.parse shared/ast/parse.ts:527:15
 INFO ide_ci::program::command: npm⚠️  ❯ new Pattern src/util/ast/match.ts:12:21
 INFO ide_ci::program::command: npm⚠️  ❯ Function.parse src/util/ast/match.ts:21:12
 INFO ide_ci::program::command: npm⚠️  ❯ src/util/ast/prefixes.ts:22:76
 INFO ide_ci::program::command: npm⚠️  ❯ Function.FromLines src/util/ast/prefixes.ts:22:39
```

<blockquote><img src="https://opengraph.githubassets.com/ccac8eaec607cf2162ebca5f580388fb142704fac5a1fd3bad1a890a029b1d6f/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Change "Override Execution Context" button to "Record" button on nodes · enso-org/enso@3f03d53</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

farmaazon commented:
> anyone know what's going on with the tests? https://github.com/enso-org/enso/actions/runs/8094866300/job/22120440156?pr=9188#step:8:466
> 
> ```
>   INFO ide_ci::program::command: npm⚠️ ⎯⎯⎯⎯⎯⎯ Failed Suites 8 ⎯⎯⎯⎯⎯⎯⎯
>  INFO ide_ci::program::command: npmℹ️ 
>  INFO ide_ci::program::command: npm⚠️ 
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/composables/__tests__/selection.test.ts [ src/composables/__tests__/selection.test.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/providers/__tests__/widgetRegistry.test.ts [ src/providers/__tests__/widgetRegistry.test.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/components/ComponentBrowser/__tests__/input.test.ts [ src/components/ComponentBrowser/__tests__/input.test.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/components/GraphEditor/__tests__/dragging.test.ts [ src/components/GraphEditor/__tests__/dragging.test.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/stores/graph/__tests__/graphDatabase.test.ts [ src/stores/graph/__tests__/graphDatabase.test.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/composables/selection.ts [ src/composables/selection.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/components/GraphEditor/collapsing.ts [ src/components/GraphEditor/collapsing.ts ]
>  INFO ide_ci::program::command: npm⚠️  FAIL  src/util/ast/node.ts [ src/util/ast/node.ts ]
>  INFO ide_ci::program::command: npm⚠️ TypeError: Cannot read properties of undefined (reading '__wbindgen_add_to_stack_pointer')
>  INFO ide_ci::program::command: npm⚠️  ❯ Module.parse rust-ffi/pkg/rust_ffi.js:156:14
>  INFO ide_ci::program::command: npm⚠️     154|         return v2;
>  INFO ide_ci::program::command: npm⚠️     155|     } finally {
>  INFO ide_ci::program::command: npm⚠️     156|         wasm.__wbindgen_add_to_stack_pointer(16);
>  INFO ide_ci::program::command: npm⚠️        |              ^
>  INFO ide_ci::program::command: npm⚠️     157|     }
>  INFO ide_ci::program::command: npm⚠️     158| }
>  INFO ide_ci::program::command: npm⚠️  ❯ parseEnso shared/ast/parse.ts:65:16
>  INFO ide_ci::program::command: npm⚠️  ❯ parseBlockWithSpans shared/ast/parse.ts:541:16
>  INFO ide_ci::program::command: npm���️  ❯ parseBlock shared/ast/parse.ts:521:10
>  INFO ide_ci::program::command: npm⚠️  ❯ Module.parse shared/ast/parse.ts:527:15
>  INFO ide_ci::program::command: npm⚠️  ❯ new Pattern src/util/ast/match.ts:12:21
>  INFO ide_ci::program::command: npm⚠️  ❯ Function.parse src/util/ast/match.ts:21:12
>  INFO ide_ci::program::command: npm⚠️  ❯ src/util/ast/prefixes.ts:22:76
>  INFO ide_ci::program::command: npm⚠️  ❯ Function.FromLines src/util/ast/prefixes.ts:22:39
> ```
> 
> > <img alt="" width="48" src="https://camo.githubusercontent.com/f34a4d98f10ee53828ba56eff2bffcbce57b97345d07113581bc67e2c6d45b22/68747470733a2f2f6f70656e67726170682e6769746875626173736574732e636f6d2f636361633865616563363037636632313632656263613566353830333838666231343237303466616335613166643362616431613839306130323962316436662f656e736f2d6f72672f656e736f"><img alt="" height="14" src="https://camo.githubusercontent.com/95e35b1ddbd18e9680ef4876f3544fdf46e2116b5fefc3f6a48e1e427ea75b59/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f66617669636f6e732f66617669636f6e2e737667"> GitHub**[Change "Override Execution Context" button to "Record" button on nodes · enso-org/enso@3f03d53](https://github.com/enso-org/enso)**Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.

Missing initializeFFI call in test file. Unfortunately, we need to add it in every test file using parser.

<blockquote><img src="https://opengraph.githubassets.com/ccac8eaec607cf2162ebca5f580388fb142704fac5a1fd3bad1a890a029b1d6f/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Change "Override Execution Context" button to "Record" button on nodes · enso-org/enso@3f03d53</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

somebody1234 committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

somebody1234 committed:
Fix tests

somebody1234 committed:
Fix tests?

somebody1234 committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

mergify[bot] committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

JaroslavTulach committed:
Use .. to identify autoscoped constructors (#9285)

farmaazon committed:
Restore whenReady, as I do not see the freeze (#9262)

When cleaning up board, I stumbled upon https://github.com/enso-org/enso/issues/5851 I tried to replace whenReady, and it just worked. Perhaps the freeze was fixed during some electron version bump.

Tests on other platforms (I'm using Garuda Linux) advised.

Fixes #5851

somebody1234 committed:
Extract tests to their own files (#9229)

Move tests to their own files, for consistency

None

somebody1234 committed:
Fix tests

somebody1234 commented:
man, whatever happened here?

somebody1234 commented:
oh i guess i messed up the merge...

somebody1234 committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

indiv0 commented:
✅ QA passed

- [x] Fix appearance of Record/Record Once icon in top menu
- [x] Make the icons be consistent with the new icons for live/design mode in https://github.com/enso-org/enso/pull/9133
    - i.e. grey circle for off, red circle for on
- [x] Unconditionally show per-node record icon if it is set
- [x] Remove the ability to override the execution context to disabled
- [x] Fix the icon for nodes with an overridden execution context always being the Enso icon

<img width="206" alt="image" src="https://github.com/enso-org/enso/assets/641547/cf93a8dd-5f34-4438-979a-1b7898d370bd">
<img width="137" alt="image" src="https://github.com/enso-org/enso/assets/641547/b7e6ccbe-fc90-4247-a9ec-eab35c7316de">
<img width="103" alt="image" src="https://github.com/enso-org/enso/assets/641547/3756bb03-3219-46bf-9fc0-aa067b8a1b02">

indiv0 commented on app/gui2/src/util/ast/node.ts:
```diff
@@ -1,5 +1,20 @@
+}
+
+/** MUST be called after `initializeFFI`. */
+export function initializePrefixes() {
+  prefixes = makePrefixes()
+}
```
If this must be called after `initializeFFI` we could encode that in the type system by making `initializeFFI` return a value that can't be constructed otherwise, and make that value an argument of this function.

But that might be overkill.

somebody1234 replied:
yeah, i'm going to consider it as overkill for now. although it wouldn't be difficult to do it, it doesn't really match up with the rest of the codebase IMO.

somebody1234 replied:
i mean, realistically in the actual app, both are called in the entrypoint in the correct order. so i don't think omitting it would be too big of a deal since there's only one place where they're called.

indiv0 reviewed (approved):
LGTM

somebody1234 committed:
Merge branch 'develop' into wip/sb/toggle-recording-on-nodes

somebody1234 committed:
Fix E2E tests
